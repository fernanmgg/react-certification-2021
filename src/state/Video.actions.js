export const actions = (dispatch) => ({
  setSearch: (payload) => dispatch({ type: 'SET_SEARCH', payload }),
  setAuth: (payload) => dispatch({ type: 'SET_AUTH', payload }),
  unsetAuth: () => dispatch({ type: 'UNSET_AUTH' }),
  addFav: (payload) => dispatch({ type: 'ADD_FAVORITE', payload }),
  remFav: (payload) => dispatch({ type: 'REMOVE_FAVORITE', payload }),
});
