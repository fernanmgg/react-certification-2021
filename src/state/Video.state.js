import React from 'react';

import { getFavorites } from '../utils/favoritesDB';

const VideoContext = React.createContext();

const initialState = {
  auth: null,
  favorites: [],
  search: 'wizeline',
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_SEARCH':
      return { ...state, search: action.payload };
    case 'SET_AUTH':
      if (action.payload.remember)
        localStorage.setItem('auth', JSON.stringify(action.payload.auth));
      else sessionStorage.setItem('auth', JSON.stringify(action.payload.auth));
      return {
        ...state,
        auth: action.payload.auth,
        favorites: getFavorites(action.payload.auth.name),
      };
    case 'UNSET_AUTH':
      localStorage.removeItem('auth');
      sessionStorage.removeItem('auth');
      return { ...state, auth: null, favorites: [] };
    default:
      throw new Error('Error: Undefined action');
  }
};

export { VideoContext, initialState, reducer };
