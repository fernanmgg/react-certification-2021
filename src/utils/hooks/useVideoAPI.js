import { useState, useEffect } from 'react';

const key = process.env.REACT_APP_YOUTUBE_API_KEY;

const baseUrl = 'https://www.googleapis.com/youtube/v3/videos?';

export default function useVideoAPI(videoId) {
  const [video, setVideo] = useState(null);
  const [redirect, setRedirect] = useState(false);
  const [fetchError, setFetchError] = useState(false);

  useEffect(() => {
    function buildURI() {
      const fields = [
        `id=${videoId}`,
        'part=snippet',
        'fields=items(snippet(title,description,thumbnails(medium(url))))',
        `key=${key}`,
      ];
      return baseUrl.concat(
        fields.reduce((previous, current) => previous.concat(`&${current}`))
      );
    }

    const fetchData = async () => {
      try {
        const response = await fetch(encodeURI(buildURI()));
        const json = await response.json();
        if (Array.isArray(json.items) && json.items.length === 1) setVideo(json.items[0]);
        else setRedirect(true);
      } catch (e) {
        setFetchError(true);
      }
    };

    const check = /^[A-Za-z0-9_-]{11}$/.test(videoId);
    if (check) fetchData();
    else setRedirect(true);
  }, [videoId]);

  return { video, redirect, fetchError };
}
