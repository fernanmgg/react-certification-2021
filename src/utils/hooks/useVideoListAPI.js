import { useState, useEffect } from 'react';

const key = process.env.REACT_APP_YOUTUBE_API_KEY;

const baseUrl = 'https://www.googleapis.com/youtube/v3/search?';

export default function useVideoListAPI(query, relatedVideos = false) {
  const [videos, setVideos] = useState([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);

  useEffect(() => {
    function buildURI() {
      const fields = [
        'part=snippet',
        'fields=items(id(videoId),snippet(title,description,thumbnails(medium(url))))',
        'maxResults=20',
        'type=video',
        `key=${key}`,
      ];
      if (!relatedVideos) fields.push(`q=${query}`);
      else fields.push(`relatedToVideoId=${query}`);
      return baseUrl.concat(
        fields.reduce((previous, current) => previous.concat(`&${current}`))
      );
    }

    const fetchData = async () => {
      setLoading(true);
      try {
        const response = await fetch(encodeURI(buildURI()));
        const json = await response.json();
        if (json.items) setVideos(json.items);
        else setError(true);
      } catch (e) {
        setError(true);
      }
      setLoading(false);
    };

    const check = !relatedVideos
      ? query.length > 0 && query.length < 64
      : /^[A-Za-z0-9_-]{11}$/.test(query);
    if (check) fetchData();
  }, [query, relatedVideos]);

  return { videos, loading, error };
}
