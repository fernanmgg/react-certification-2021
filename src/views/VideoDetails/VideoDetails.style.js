import styled from 'styled-components';

const StyledVideoDetails = styled.div`
  background-color: ${(props) => props.theme.backgroundTransparent};
  margin-top: 16px;
  padding: 16px;
  width: 100%;
  display: flex;
  flex-direction: row;
`;

const Wrapper = styled.div`
  flex: 3 1 0;
  margin: 0 16px;
`;

const Title = styled.h1`
  color: ${(props) => props.theme.text};
`;

const Description = styled.h3`
  color: ${(props) => props.theme.textLight};
`;

export { StyledVideoDetails, Wrapper, Title, Description };
