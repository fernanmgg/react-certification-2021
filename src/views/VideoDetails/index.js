export { default } from './VideoDetails.view';
