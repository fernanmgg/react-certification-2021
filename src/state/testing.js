import React from 'react';
import { ThemeProvider } from 'styled-components';

import { VideoContext } from './Video.state';
import { lightTheme as theme } from '../components/App/App.theme';

const wrapWithThemeContext = (children) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const wrapWithVideoContext = (children, state = {}, actions = {}) => {
  return (
    <VideoContext.Provider value={{ state, ...actions }}>
      {children}
    </VideoContext.Provider>
  );
};

export { wrapWithThemeContext, wrapWithVideoContext };
