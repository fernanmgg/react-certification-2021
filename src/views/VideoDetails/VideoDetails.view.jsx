import React from 'react';
import { useParams } from 'react-router-dom';

import Details from '../../components/Details';
import useVideoListAPI from '../../utils/hooks/useVideoListAPI';

function VideoDetails() {
  const { videoId } = useParams();
  const { videos, loading, error } = useVideoListAPI(videoId, true);
  return <Details videos={videos} loading={loading} error={error} />;
}

export default VideoDetails;
