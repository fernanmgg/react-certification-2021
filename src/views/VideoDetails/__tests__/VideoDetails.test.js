import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router';
import 'jest-styled-components';

import VideoDetails from '../VideoDetails.view';
import * as useVideoAPI from '../../../utils/hooks/useVideoAPI';
import * as useVideoListAPI from '../../../utils/hooks/useVideoListAPI';
import { wrapWithVideoContext } from '../../../state/testing';

describe('VideoDetails UI tests', () => {
  test('renders correctly', () => {
    const video = {
      snippet: {
        title: 'test title',
        description: 'test description',
        thumbnails: {
          medium: {
            url: 'test',
          },
        },
      },
    };
    useVideoAPI.default = jest.fn(() => ({ video, redirect: false, fetchError: false }));
    useVideoListAPI.default = jest.fn(() => ({
      videos: [],
      loading: false,
      error: false,
    }));
    const { asFragment } = render(
      wrapWithVideoContext(
        <MemoryRouter initialEntries={['/video/testId']}>
          <Route path="/video/:videoId">
            <VideoDetails />
          </Route>
        </MemoryRouter>,
        { auth: null },
        jest.fn()
      )
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
