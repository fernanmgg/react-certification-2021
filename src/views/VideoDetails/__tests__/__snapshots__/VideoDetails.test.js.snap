// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetails UI tests renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: calc(100vh - 64px - 16px);
  margin-top: 16px;
  padding: 16px;
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 3 1 0;
  -ms-flex: 3 1 0;
  flex: 3 1 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  margin: 0 16px;
}

.c2 {
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  font-size: 1.5rem;
}

.c5 {
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  font-size: 1rem;
  overflow-y: auto;
  white-space: pre-wrap;
}

.c6 {
  border-radius: 8px;
  -webkit-flex: 1 1 0;
  -ms-flex: 1 1 0;
  flex: 1 1 0;
  overflow-y: scroll;
  padding-top: 8px;
}

.c7 {
  text-align: center;
  width: 100%;
}

@media (max-width:900px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    height: 200vh;
  }
}

@media (max-width:900px) {
  .c1 {
    margin: 16px 0;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <iframe
          height="100%"
          src="https://www.youtube.com/embed/testId"
          title="testId"
          width="100%"
        />
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          test title
        </div>
        <div
          class="c5"
        >
          test description
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <h2
        class="c7"
      >
        No videos found...
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
